### PR TITLE
[PySpark] - Allow union all by name

### PR DESCRIPTION
**Pull request by mariotaddeucci**

### CODE REVIEW

File: tools/pythonpkg/duckdb/experimental/spark/sql/dataframe.py [hunk 1]
```diff
@@ -846,9 +846,23 @@ def unionByName(
...
-        raise NotImplementedError
+        if allowMissingColumns:
+            for col in set(self.columns) - set(other.columns):
+                other = other.withColumn(
+                    col,
+                    Column(ConstantExpression(None)).cast(self.schema[col].dataType),
+                )
+
+            for col in set(other.columns) - set(self.columns):
+                self = self.withColumn(
```
Comment (Tishj) [inline tools/pythonpkg/duckdb/experimental/spark/sql/dataframe.py:857]: This is altering self which I don't think we want to do, this should be a "const" method with no side effects beyond the returned result

File: tools/pythonpkg/duckdb/experimental/spark/sql/dataframe.py [hunk 2]
```diff
@@ -846,9 +846,23 @@ def unionByName(
         |NULL|   4|   5|   6|
         +----+----+----+----+
         """
-        if not allowMissingColumns:
-            raise ContributionsAcceptedError
-        raise NotImplementedError
+        if allowMissingColumns:
+            for col in set(self.columns) - set(other.columns):
```
Comment (Tishj) [inline tools/pythonpkg/duckdb/experimental/spark/sql/dataframe.py:850]: I also wonder if this has side effects on the passed in `other`, I suspect it does
It shouldn't.

**Comment by mariotaddeucci** [comment]
@Tishj fixed, no side effects

**Review by Tishj — CHANGES_REQUESTED** [review]
Thanks, could we add more tests?
- testing the `"Union by name requires the same number of columns"` error
- missing columns on the `self` side (allowMissingColumns)
- missing columns on the `other` side (allowMissingColumns)

File: tools/pythonpkg/tests/fast/spark/test_spark_union_by_name.py [hunk 3]
```diff
@@ -32,10 +17,34 @@ def df2(spark):
...
-        print(res)
+
+        assert rel.columns == ["name", "id"]
+        assert rel.count() == 8
+
+        assert rel.select("name").distinct().count() == 6
+        assert rel.select("id").distinct().count() == 5
+
+    def test_error_union_by_name(self, df1, df3):
+        with pytest.raises(ValueError):
```
Comment (Tishj) [inline tools/pythonpkg/tests/fast/spark/test_spark_union_by_name.py:40]: please use `match: ...` to check for an explicit error, ValueError is too generic

**Comment by mariotaddeucci** [comment]
@Mytherin is there something to change?

**Comment by Mytherin** [comment]
There's still some open comments from @Tishj, as well as a merge conflict. Could you have another look?

**Pull request closed without merging**